### PR TITLE
Include anonymous categories when evaluating cursor entities.

### LINKED ISSUE #8: Doesn't seem to take into consideration categories
In the following before/after situation, objc-diff seems to believe that `someMethod` is deleted rather than deprecated. My guess is that the diff engine may not be considering APIs across categories.

Before:

```
@interface MyController : UIViewController
@end

@interface MyController ()

- (SomeObject *)someMethod;

@end
```

After:

```
@interface MyController : UIViewController

- (SomeObject *)otherMethod;

@end

@interface MyController ()

- (SomeObject *)someMethod
__deprecated_msg("Use otherMethod instead.");

@end
```

### PR DESCRIPTION
**Pull request by jverkoey**
This closes https://github.com/mattstevens/objc-diff/issues/8. I've verified that tests pass and that my local invocation generates the expected output now.

### CODE REVIEW

File: OCDiffTests/OCDAPIComparatorTests.m [hunk 1]
```diff
@@ -1171,7 +1189,7 @@ - (void)testAddRemoveForName:(NSString *)name base:(NSString *)base addition:(NS
     XCTAssertEqualObjects(differences, @[], @"Unchanged test failed for %@", name);
 
     // Unchanged, different line number
-    differences = [self differencesBetweenOldSource:addition newSource:[@"\n" stringByAppendingString:addition]];
+    differences = [self differencesBetweenOldSource:addition newSource:[@"" stringByAppendingString:addition]];
```
Comment (mattstevens) [inline OCDiffTests/OCDAPIComparatorTests.m]: The newline is an essential part of this test so this needs to be put back. Where possible objc-diff uses the Clang-defined Unified Symbol Resolution (USR) as the key on which to determine if an API has been added or removed. For declarations that are not externally visible Clang includes location information in the USR if the declaration is not in a system header (system headers are considered to be sane). Unfortunately this means that the USR cannot be used as a key for these entity types as simply moving them to a different line would look like a removal an addition to objc-diff. For such cases [`keyForCursor:`](https://github.com/mattstevens/objc-diff/blob/95998186066dc66e50e291605d011d8b6a4d4d52/OCDiff/OCDAPIComparator.m#L147) falls back to generating and objc-diff-specific key.

**Comment by mattstevens** [comment]
Thanks, and thanks for adding tests! I had one important comment, the rest looks good to me.

File: OCDiffTests/OCDAPIComparatorTests.m (hunk 1 above)
Comment (jverkoey) [inline OCDiffTests/OCDAPIComparatorTests.m]: Ahh, this helps explain why the test was failing with this newline. I'll revisit.

**Comment by jverkoey** [comment]
Off for the weekend, will post back here when I get a chance to look at this again! Thanks again for the guidance.

**Comment by jverkoey** [comment]
I've changed the code to better support USRs with line numbers in them. My simple tests now pass.

My approach was to flatten the USRs into a fuzzy variant where the line numbers are removed.

My gut is saying that this will fail in situations where there exist multiple anonymous categories in a file. I'm going to write tests for that now and report back.

**Comment by jverkoey** [comment]
I'm now looking into an approach where I enumerate the APIs, search for fuzzy-duplicates, sort them based on their line number and give them unique relatively-indexed numbers instead.

**Comment by jverkoey** [comment]
New tests now include multiple anonymous categories. New code passes all tests.

Next test to try is reordering two anonymous categories. I expect this to fail but this may be "the best we can do" at this point. Detecting reordered anonymous categories is likely non-trivial follow-up work.

**Comment by jverkoey** [comment]
I've confirmed that the latest changes work on my code as well. I think this is in a state to be looked at again. Let me know what you think :)

**Comment by jverkoey** [comment]
I've also confirmed with a test that reordering anonymous categories behaves as expected: it produces a diff for both categories.

**Comment by mattstevens** [comment]
I was thinking about this one today and decided to handle this issue in a different way. Rather than reporting class extensions (anonymous categories) I opted to exclude them but include their children, then report the extensions that the class extension makes to the class as modifications to the class itself. For example, if a class extension declares conformance to an additional protocol, this is reported as a modification to the list of conformed-to protocols for the class. I think this better reflects the affect of a class extension on the API in the report, and it avoids the need to deal with reporting entries for multiple class extensions consistently.

**Pull request closed without merging**